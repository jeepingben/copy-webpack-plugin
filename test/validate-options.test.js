import CopyPlugin from '../src/index';

// Todo remove after dorp node@6 support
if (!Object.entries) {
  Object.entries = function entries(obj) {
    const ownProps = Object.keys(obj);
    let i = ownProps.length;
    const resArray = new Array(i);

    // eslint-disable-next-line no-plusplus
    while (i--) {
      resArray[i] = [ownProps[i], obj[ownProps[i]]];
    }

    return resArray;
  };
}

describe('validate options', () => {
  const tests = {
    patterns: {
      success: [
        ['test.txt'],
        ['test.txt', 'test-other.txt'],
        [
          'test.txt',
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
          },
        ],
        [
          {
            from: 'test.txt',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
          },
        ],
        [
          {
            from: 'test.txt',
            context: 'context',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            toType: 'file',
            test: /test/,
            force: true,
            ignore: [
              'ignore-1',
              'ignore-2',
              {
                dot: false,
                glob: '**/*',
              },
            ],
            flatten: true,
            cache: true,
            transform: () => {},
            transformPath: () => {},
          },
        ],
        [
          {
            from: {
              glob: '**/*',
              dot: false,
            },
            to: 'dir',
            context: 'context',
            globOptions: {
              dot: false,
            },
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            globOptions: {
              dot: false,
            },
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            test: 'test',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            test: /test/,
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            cache: {
              foo: 'bar',
            },
          },
        ],
      ],
      failure: [
        true,
        'true',
        '',
        {},
        [''],
        [{}],
        [
          {
            from: '',
            to: 'dir',
            context: 'context',
          },
        ],
        [
          {
            from: true,
            to: 'dir',
            context: 'context',
          },
        ],
        [
          {
            from: 'test.txt',
            to: true,
            context: 'context',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: true,
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            toType: 'foo',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            test: true,
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            force: 'true',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            ignore: true,
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            ignore: [true],
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            ignore: ['test.txt', true],
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            flatten: 'true',
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            cache: () => {},
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            transform: true,
          },
        ],
        [
          {
            from: 'test.txt',
            to: 'dir',
            context: 'context',
            transformPath: true,
          },
        ],
      ],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      let error;

      try {
        // eslint-disable-next-line no-new
        new CopyPlugin(key === 'patterns' ? value : { [key]: value });
      } catch (errorFromPlugin) {
        if (errorFromPlugin.name !== 'ValidationError') {
          throw errorFromPlugin;
        }

        error = errorFromPlugin;
      } finally {
        if (type === 'success') {
          expect(error).toBeUndefined();
        } else if (type === 'failure') {
          expect(() => {
            throw error;
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
