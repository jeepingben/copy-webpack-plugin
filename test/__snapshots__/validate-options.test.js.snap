// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "patterns" option with "" value 1`] = `
"CopyPlugin Invalid Options

options should be array
"
`;

exports[`validate options should throw an error on the "patterns" option with "[""]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should NOT be shorter than 1 characters
options.0 should be object
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"","to":"dir","context":"context"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.from should NOT be shorter than 1 characters
options.0.from should be object
options.0.from should match some schema in anyOf
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","flatten":"true"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.flatten should be boolean
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","force":"true"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.force should be boolean
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","ignore":["test.txt",true]}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.ignore.1 should be string
options.0.ignore.1 should be object
options.0.ignore.1 should match some schema in anyOf
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","ignore":[true]}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.ignore.0 should be string
options.0.ignore.0 should be object
options.0.ignore.0 should match some schema in anyOf
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","ignore":true}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.ignore should be array
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","test":true}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.test should be string
options.0.test should pass \\"instanceof\\" keyword validation
options.0.test should match some schema in anyOf
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","toType":"foo"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.toType should be equal to one of the allowed values
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","transform":true}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.transform should pass \\"instanceof\\" keyword validation
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context","transformPath":true}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.transformPath should pass \\"instanceof\\" keyword validation
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":"context"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.cache should be boolean
options.0.cache should be object
options.0.cache should match some schema in anyOf
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":"dir","context":true}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.context should be string
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":"test.txt","to":true,"context":"context"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.to should be string
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{"from":true,"to":"dir","context":"context"}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0.from should be string
options.0.from should be object
options.0.from should match some schema in anyOf
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "[{}]" value 1`] = `
"CopyPlugin Invalid Options

options.0 should be string
options.0 should have required property 'from'
options.0 should match some schema in anyOf
"
`;

exports[`validate options should throw an error on the "patterns" option with "{}" value 1`] = `
"CopyPlugin Invalid Options

options should be array
"
`;

exports[`validate options should throw an error on the "patterns" option with "true" value 1`] = `
"CopyPlugin Invalid Options

options should be array
"
`;

exports[`validate options should throw an error on the "patterns" option with "true" value 2`] = `
"CopyPlugin Invalid Options

options should be array
"
`;
